rclpy/src/rclpy/_rclpy_qos_event.c

```diff
@@ -0,0 +1,357 @@
+  * \param[in] pyparent Capsule containing the parent Publisher or Subscription.
+  * \return capsule containing rcl_event_t.
+  * \return NULL on failure.
+  */
+static PyObject *
+rclpy_create_event(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  unsigned PY_LONG_LONG event_type;
+  PyObject * pyparent = NULL;
+
+  rcl_ret_t ret;
+  rcl_subscription_t * subscription = NULL;
+  rcl_publisher_t * publisher = NULL;
+  rcl_event_t * event = NULL;
+
+  PyObject * pyevent = NULL;
+
+  if (!PyArg_ParseTuple(args, "KO", &event_type, &pyparent)) {
+    return NULL;
+  }
+
+  if (_is_pycapsule_rcl_subscription(pyparent)) {
+    subscription = _pycapsule_to_rcl_subscription(pyparent);
+  } else if (_is_pycapsule_rcl_publisher(pyparent)) {
+    publisher = _pycapsule_to_rcl_publisher(pyparent);
+  } else {
+    PyErr_Format(PyExc_TypeError, "Event parent was not a valid Publisher or Subscription.");
+    return NULL;
+  }
+
+  event = _new_zero_initialized_rcl_event();
+  if (!event) {
+    return NULL;
+  }
+
+  if (subscription) {
+    ret = rcl_subscription_event_init(event, subscription, event_type);
+  } else {
+    ret = rcl_publisher_event_init(event, publisher, event_type);
+  }
+  if (!_check_rcl_return(ret, "Failed to initialize event")) {
+    PyMem_Free(event);
+    return NULL;
+  }
+
+  pyevent = PyCapsule_New(event, "rcl_event_t", _destroy_event_capsule);
+  if (!pyevent) {
+    ret = rcl_event_fini(event);
+    PyMem_Free(event);
+    _check_rcl_return(ret, "Failed to fini 'rcl_event_t'");
+    return NULL;
+  }
+
+  return pyevent;
+}
+
+/// Get a pending QoS event's data.
+/**
+  * After having determined that a middleware event is ready, get the callback payload.
+  *
+  * Raises RuntimeError on failure to take the event from the middleware.
+  * Raises TypeError if the capsules are not the correct types.
+  * Raises ValueError on unknown event_type argument.
+  *
+  * \param[in] pyevent Event handle from rclpy_create_event.
+  * \param[in] event_type Enum value of
+  *   rcl_publisher_event_type_t or rcl_subscription_event_type_t, chosen by the type of pyparent.
+  * \param[in] pyparent Capsule containing the parent Publisher or Subscription.
+  * \return Python object from rclpy.qos_event containing callback data.
+  * \return NULL on failure.
+  */
+static PyObject *
+rclpy_take_event(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  // Arguments
+  PyObject * pyevent = NULL;
+  PyObject * pyparent = NULL;
+  unsigned PY_LONG_LONG event_type;
+
+  // Type conversion
+  rcl_ret_t ret;
+  rcl_event_t * event = NULL;
+  _qos_event_callback_data_t event_data;
+  _qos_event_data_filler_function event_filler = NULL;
+
+  if (!PyArg_ParseTuple(args, "OOK", &pyevent, &pyparent, &event_type)) {
+    return NULL;
+  }
+
+  event = _pycapsule_to_rcl_event(pyevent);
+  if (!event) {
+    return NULL;
+  }
+
+  event_filler = _get_qos_event_data_filler_function_for(pyparent, event_type);
```

Comment:
Mind looking at the windows warning? https://ci.ros2.org/job/ci_windows/6775/warnings43Result/new/

Reply:
That was fast, I assume this solves it 107baa6d7349f179a0cd789e11d7194c653eeb25

Reply:
Yeah - I just pushed a commit but unfortunately I don't have a windows dev env set up right now. It doesn't introduce warnings on linux (still building mac - it's a lot slower)

Reply:
@sloretz mac build looks good, this change also does not introduce any warnings there - should we just kick off another build? i've updated the gist to use the master version of everything except this, since the dependencies merged

https://gist.githubusercontent.com/emersonknapp/67d93f9e90183987832715b5979640e8/raw/7eb11515e048841d603d35ce0a5e6fecfdf9296f/qos-4-callbacks.repos